ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,190 @@
+	fn initial(&self) -> Option<String> {
+		// initial payload may contain additional data, not in this case
+		self.job()
+	}
+
+	fn job(&self) -> Option<String> {
+		{
+			let last_work = self.last_work.write();
+			if let Some((pow_hash, difficulty, number)) = *last_work {
+				return Some(self.payload(pow_hash, difficulty, number));
+			}
+		}
+
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		if let Some((pow_hash, difficulty, number)) = miner.map_sealing_work(&*client, |b| {
+			let pow_hash = b.hash();
+			let number = b.block().header().number();
+			let difficulty = b.block().header().difficulty();
+
+			(pow_hash, *difficulty, number)
+		}) {
+			*self.last_work.write() = Some((pow_hash, difficulty, number));
+			Some(self.payload(pow_hash, difficulty, number))
+		} else { None }
+	}
+
+	fn submit(&self, payload: Vec<String>) {
+		if payload.len() != 3 {
+			warn!(target: "stratum", "submit_work: invalid work submit request({:?})", payload);
+		}
+
+		let nonce = match H64::from_str(&payload[0]) {
+			Ok(nonce) => nonce,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid nonce ({:?})", e);
+				return;
+			}
+		};
+
+
+		let pow_hash = match H256::from_str(&payload[1]) {
+			Ok(pow_hash) => pow_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid hash ({:?})", e);
+				return;
+			}
+		};
+
+		let mix_hash = match H256::from_str(&payload[2]) {
+			Ok(mix_hash) => mix_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid mix-hash ({:?})", e);
+				return;
+			}
+		};
+
+		trace!(target: "stratum", "submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}", nonce, pow_hash, mix_hash);
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		let seal = vec![encode(&mix_hash).to_vec(), encode(&nonce).to_vec()];
+		if let Err(e) = miner.submit_seal(&*client, pow_hash, seal) {
+			warn!(target: "stratum", "submit_work error: {:?}", e);
+		};
+	}
+}
+
+impl StratumJobDispatcher {
+	/// New stratum job dispatcher given the miner and client
+	fn new(miner: &Arc<Miner>, client: &Arc<Client>) -> StratumJobDispatcher {
+		StratumJobDispatcher {
+			seed_compute: Mutex::new(SeedHashCompute::new()),
+			last_work: RwLock::new(None),
+			client: Arc::downgrade(client),
+			miner: Arc::downgrade(miner),
+		}
+	}
+
+	/// Serializes payload for stratum service
+	fn payload(&self, pow_hash: H256, difficulty: U256, number: u64) -> String {
+		// TODO: move this to engine
+		let target = Ethash::difficulty_to_boundary(&difficulty);
+		let seed_hash = &self.seed_compute.lock().get_seedhash(number);
+		let seed_hash = H256::from_slice(&seed_hash[..]);
+		format!(
+			r#"["0x{}","0x{}","0x{}","0x{:x}"]"#,
+			pow_hash.hex(), seed_hash.hex(), target.hex(), number
+		)
+	}
+}
+
+/// Wrapper for dedicated stratum service
+pub struct Stratum {
+	dispatcher: Arc<StratumJobDispatcher>,
+	base_dir: String,
```

Comment:
it's not path, it's socket address in general
can be modified to tcp/ip later
and is not subject to `/` or `\` separator issues which is pathbuf created to take care of